### PR TITLE
Implement `fmemopen`

### LINKED ISSUE #527: posix: fmemopen() unimplemented
Mount from util-linux requires that to work.

### PR DESCRIPTION
**Pull request by no92**
Supersedes #540.
Fixes #527.

### CODE REVIEW

**Comment by no92** [comment]
The failing assert will not be done on glibc due to (incorrectly) failing there. NetBSD's libc tests also guard against this in the same way. My implementation is adhering to POSIX, and thus correct; musl and the BSDs agree with me here. This should be ready for review now.

**Review by ikbenlike — APPROVED** [review]
LGTM

**Review by Geertiebear — COMMENTED** [review]
Haven't taken a good look at this yet, will do later, but just a nitpick already when skimming it over: the comments aren't in the same style :) In general we follow that comments should ``// Look like this.``, if you could fix that in the relevant places that would be really nice.

**Comment by no92** [comment]
Fixed the comments.

File: options/posix/generic/posix-file-io.cpp [hunk 1]
```diff
@@ -6,12 +6,39 @@
...
+: abstract_file{do_dispose}, _inBuffer{in_buf}, _inBufferSize{size}, _flags{flags},
+	_allowResizing{false} {
+	if(!_inBuffer) {
+		_inBuffer = getAllocator().allocate(size);
+		_needsDeallocation = true;
+	}
+
+	if(_flags & O_APPEND) {
+		// the initial seek-size for append is zero if buf was NULL, or the first '\0' found, or the size
+		_max_size = (_needsDeallocation) ? 0 : strnlen(reinterpret_cast<char *>(_inBuffer), _inBufferSize);
```
Comment (Geertiebear) [inline options/posix/generic/posix-file-io.cpp:25]: While this is correct, there is missing functionality here, which is that append should also set the position of the stream to the first null byte. This functionality is also not tested in the tests afaict, which is why that hasn't been caught.

File: tests/posix/fmemopen.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	f = fmemopen(NULL, BUFFER0_SIZE, "a");
+	assert(f);
+
+	assert(!fseek(f, 0, SEEK_END));
+	pos = ftell(f);
+	assert(pos == 0);
+
+	fclose(f);
+
+	// test seek with mode "a" and a buffer containing a '\0'
```
Comment (Geertiebear) [inline tests/posix/fmemopen.c:56]: Missing test case: testing the initial state of the buffer as mandated by the spec.

File: tests/posix/fmemopen.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	// > use of fseek() or fseeko() on the stream with SEEK_END shall seek relative to this size.
+	// > For modes r and r+ the size shall be set to the value given by the size argument.
+#if !defined(__GLIBC__)
+	assert(pos == BUFFER0_SIZE);
+#endif
+
+	fclose(f);
+
+	// test seek with mode "w"
+	f = fmemopen(NULL, BUFFER0_SIZE, "w");
```
Comment (Geertiebear) [inline tests/posix/fmemopen.c:37]: Missing test case: testing the initial conditions of the buffer (the implementation should place a null byte at the first byte).

File: options/posix/include/mlibc/posix-file-io.hpp [hunk 4]
```diff
@@ -21,13 +23,42 @@ struct mem_file : abstract_file {
 	int io_seek(off_t offset, int whence, off_t *new_offset) override;
 private:
 	void _update_ptrs();
+
+	// Convenience function to obtain the correct buffer
+	frg::span<char> _buffer() {
```
Comment (Geertiebear) [inline options/posix/include/mlibc/posix-file-io.hpp:28]: Both of these functions could and should be marked ``const``.

**Review by Geertiebear — CHANGES_REQUESTED** [review]
In general, I'm not a fan of this code. It seems that the ``_allowResizing`` flag was used to cram two different classes into one implementation, which makes for a very confusing read of the code. In this case it would be appropriate to have two different classes, and possibly a common superclass that abstracts some of the common behaviour away. Right now (what should be) virtual calls are emulated using the ``_allowResizing`` flag.

In summary, please add missing test cases, since there were several untested cases I've noticed. And separate the two different implementations using inheritance, this will clean up the implementation and hopefully reduce the likelihood of bugs.

File: tests/posix/fmemopen.c (hunk 3 above)
Comment (Geertiebear) [reply tests/posix/fmemopen.c:37]: Yeah that's fair. I think we don't have to test for this.

**Review by Geertiebear — COMMENTED** [review]
I feel like ``io_seek()`` and ``io_read()`` can be fully abstracted by just making ``_buffer_size()`` a protected virtual function in the base class. ``io_write`` is unfortunately a little bit trickier it seems since there is a special case for ``fmemopen()``.

File: options/posix/include/mlibc/posix-file-io.hpp (hunk 4 above)
Comment (Geertiebear) [reply options/posix/include/mlibc/posix-file-io.hpp]: You can still mark the function as ``const`` just not the return value

File: options/posix/generic/posix-file-io.cpp [hunk 5]
```diff
@@ -30,58 +21,160 @@ int mem_file::determine_bufmode(buffer_mode *mode) {
...
-void mem_file::_update_ptrs() {
+void memstream_mem_file::_update_ptrs() {
 	*_bufloc = _buf.data();
 	*_sizeloc = _buf.size() - 1;
 }
 
+fmemopen_mem_file::fmemopen_mem_file(void *in_buf, size_t size, int flags, void (*do_dispose)(abstract_file *))
+: mem_file{flags, do_dispose}, _inBuffer{in_buf}, _inBufferSize{size} {
+	if(!_inBuffer) {
+		_inBuffer = getAllocator().allocate(size);
```
Comment (Geertiebear) [inline options/posix/generic/posix-file-io.cpp:100]: Do we properly handle the case when ``size`` is 0? On glibc this will cause the first read to return EOF, but im not sure that happens here (don't we just crash?) This behavior should obviously be tested ofc.

Comment (no92) [reply options/posix/generic/posix-file-io.cpp:100]: We correctly implement it, although we could also return EINVAL and still be compliant with POSIX. I added a test case.

Comment (no92) [reply options/posix/generic/posix-file-io.cpp:100]: Fun side note: GHA's glibc actually returns EINVAL with size 0.
